### ISSUE TITLE
Vulnerability Report: model load vulnerable to arbitrary code execution via untrusted file deserialization

### ISSUE THREAD

**Issue by ybdesire** [xorbitsai/inference#3190]
### Description

The provided Python code snippet below at latest master branch is vulnerable to the CWE - 502: Deserialization of Untrusted Data vulnerability. This vulnerability stems from the use of the `torch.load` function without proper security measures. When `torch.load` is used to deserialize data from an untrusted source, it can execute arbitrary code during the deserialization process. 

https://github.com/xorbitsai/inference/blob/main/xinference/thirdparty/cosyvoice/cli/model.py#L61C1-L72C1

In this code, the `load` method attempts to load model state dictionaries from three different model files (`llm_model`, `flow_model`, and `hift_model`) using `torch.load`. However, it does not specify the `weights_only=True` parameter. This means that if any of these files contain malicious pickle data, the code will execute the arbitrary code embedded in the data, potentially leading to serious security risks such as unauthorized access, data leakage, or system compromise.

The vulnerability is further exacerbated as the code does not perform any input validation or sanitization on the file paths (`llm_model`, `flow_model`, `hift_model`), making it easier for an attacker to supply malicious files. 

This vulnerability is not directly related to the system or software version. It exists due to the improper use of the torch.load function in the Python code.

### Running Xinference with Docker? / 是否使用 Docker 运行 Xinfernece？

- [ ] docker / docker
- [x] pip install / 通过 pip install 安装
- [ ] installation from source / 从源码安装

### Version info / 版本信息

This vulnerability is not directly related to the system or software version. It exists due to the improper use of the torch.load function in the Python code.

### The command used to start Xinference / 用以启动 xinference 的命令

This vulnerability is not directly related to the system or software version. It exists due to the improper use of the torch.load function in the Python code.

#### Prerequisites
1. A Python environment with PyTorch installed.
2. A set of malicious pickle files that can be used to demonstrate the vulnerability. For the purpose of this reproduction, we can assume that an attacker has crafted these files with malicious code.

#### Steps

1. **Prepare the malicious files**:
   - An attacker creates malicious pickle files named `malicious_llm_model.pth`, `malicious_flow_model.pth`, and `malicious_hift_model.pth`. These files contain arbitrary code that will be executed during deserialization.

2. **Modify the code to use the malicious files**:
   - Replace the `llm_model`, `flow_model`, and `hift_model` arguments in the `load` method call with the paths to the malicious files. 

3. **Execute the code**:
   - Run the Python script containing the modified `load` method. When the `torch.load` function deserializes the malicious pickle files, the arbitrary code embedded in the files will be executed, demonstrating the vulnerability.

To mitigate this vulnerability, the `weights_only=True` parameter should be added to all `torch.load` calls in the `load` method. For example:
```python
self.llm.load_state_dict(torch.load(llm_model, map_location=self.device, weights_only=True), strict=True)
```

### Expected behavior / 期待表现

When the `load` method is executed, it should safely load the legitimate model state dictionaries from the provided files (`llm_model`, `flow_model`, and `hift_model`). It is expected that the `torch.load` function only loads the model weights without executing any malicious code. 

The model state dictionaries should be loaded correctly into the respective models (`self.llm`, `self.flow`, and `self.hift`), and the models should be properly configured to the specified device (`self.device`). If the `self.fp16` flag is set to `True`, the `self.llm` model should be converted to half - precision. All models should then be set to evaluation mode. 

Throughout this process, there should be no execution of unauthorized code, and the system should remain secure, without any risk of data leakage, unauthorized access, or system compromise.

**Comment by ybdesire** [comment]
The vulnerability impacted the release from [v0.15.0](https://github.com/xorbitsai/inference/releases/tag/v0.15.0) to latest [v1.4.1](https://github.com/xorbitsai/inference/releases/tag/v1.4.1)

**Comment by qinxuye** [comment]
This is inherited from cosyvoice code, do you have any comment about solving this vulnerability?

**Comment by ybdesire** [comment]
@qinxuye To mitigate this vulnerability, the weights_only=True parameter should be added to all torch.load calls in the load method. For example:
```python
self.llm.load_state_dict(torch.load(llm_model, map_location=self.device, weights_only=True), strict=True)
```

**Comment by qinxuye** [comment]
Thanks, do you have interest to send a PR?

**Comment by ybdesire** [comment]
> Thanks, do you have interest to send a PR?

I will do more research for fixing.
